### ISSUE TITLE
Asking for `workspace/configuration` on every request seems too much (and is probably racy)

### ISSUE THREAD

**Issue by krobelus** [latex-lsp/texlab#533]
[Sorry, wall of text ahead, and I might have a solution for my original problem, but some of this could be useful for texlab]

I'm debugging an issue in [kak-lsp](https://github.com/kak-lsp/kak-lsp) and I noticed something weird.
I'll first describe a somewhat surprising behavior of texlab, and then my original issue (which may not be relevant).

`server::formatting()` calls
`server::handle_feature_request()` which calls
`server::feature_request()`, which does this:

	```rs
	self.pool.execute(move || {
	    let client_capabilities = &cx.client_capabilities.lock().unwrap().clone();
	    pull_config(&req_queue, &sender, &cx.options, &client_capabilities);
	});
	```
where `pull_config` requests `workspace/configuration` in the background, and updates our context's options (atomically, via a RwLock)

later on we'll run `features::formatting::format_source_code()` which depends on the context options.

This looks like a race - we might get different results, depending on how fast `workspace/configuration` is answered.
Of course this may not be a problem in practice, because it is safe to say that the request roundtrip will always be slower than anything texlab might need to do to prepare the actual request.
(Which means that any configuration changes will only be visible to the *next* request, but let's leave that aside for now)

So this seems benign, but let me explain the problem kak-lsp is facing.

Running `:lsp-formatting-sync` command (which requests `textDocument/formatting`) on a latex document will block the editor (Kakoune) until the server sends a response.
The blocking is a feature, to avoid concurrent buffer modification in non-interactive scenarios (there is also an asynchronous version, `:lsp-formatting`).

Here's a log excerpt of what happpens on `:lsp-formatting-sync`

	Dec 31 16:03:42.109 DEBG To server: {"jsonrpc":"2.0","method":"textDocument/formatting","params":{"options":{"insertSpaces":true,"tabSize":8},"textDocument":{"uri":"file:///home/johannes/git/kak-lsp/t/latex/a.tex"}},"id":1}, module: kak_lsp::language_server_transport:177
	Dec 31 16:03:42.109 DEBG From server: {"jsonrpc":"2.0","id":1,"method":"workspace/configuration","params":{"items":[{"section":"texlab"}]}}, module: kak_lsp::language_server_transport:151
	Dec 31 16:03:42.110 DEBG To editor `1695768`: lsp-get-config '/tmp/kak-lsp/johannes/653a089e76673d5b', module: kak_lsp::editor_transport:84
	Dec 31 16:03:42.281 DEBG From server: {"jsonrpc":"2.0","id":1,"result":[{"newText":"the new text","range":{"end":{"character":0,"line":311},"start":{"character":0,"line":0}}}]}, module: kak_lsp::language_server_transport:151
	[Kakoune blocks forever]

Here kak-lsp requests `textDocument/formatting` whereupon texlab requests `workspace/configuration` (the fact that the id params are the same is pure coincidence).
In line 3, kak-lsp asks the editor for a response to `workspace/configuration` via `lsp-get-config`.
We won't ever get it, because Kakoune is still blocked, waiting for the formatting response. It's a deadlock :(

Let me summarize some problems + potential solutions
- as mentioned above, pulling workspace/configuration without synchronization is racy
- this is not really important to change
- it could be changed by first waiting for workspace/configuration to finish, and only then send the actual request
- *however*, that won't fix the kak-lsp problem, which is essentially that kak-lsp can't handle certain requests concurrently.
- I recommend just not asking for `workspace/configuration` on every request. Instead, do what other servers do: ask for it when opening a file (`textDocument/didOpen`), and perhaps when the server sends `workspace/didChangeConfiguration` (IIRC that also contains payload you could use, but the lazy version is to just ignore the payload and respond with a request for `workspace/configuration`).
  - (the details are from my memory, not 100% accurate. I don't think it's specified anywhere when to send `workspace/configuration` but I usually just check what gopls does)
- Yes, this is definitely a bug on the kak-lsp side, but I think my suggested change makes things simpler overall. Blocking the editor's UI thread is the simplest way to implement synchronization, and many text editors are single-threaded programs.
- After typing this, I realize that there is a fairly simple solution on the kak-lsp side: We can prepare the workspace/configuration response eagerly  (e.g. run `lsp-get-config` whenever that config changes). Trying that now.

**Comment by krobelus** [comment]
Ok I fixed the bug in kak-lsp with https://github.com/kak-lsp/kak-lsp/commit/a12831a1a3d755d1ef9400447979010e685f1061 but maybe it's still worth rethinking workspace/configuration

**Comment by pfoerster** [comment]
Thanks for the report. Reducing the number of `workspace/configuration` is definitely a good idea. In the past, we also had problems with Emacs (https://github.com/latex-lsp/texlab/issues/456), where some requests were not answered. Regarding `workspace/didChangeConfiguration`: I think the correct behavior is now to assume that the settings are empty and just query the configuration whenever the notification is sent. This should be a lot better than spamming the client all the time.

**Comment by pfoerster** [comment]
@krobelus
The spamming of `workspace/configuration` should be fixed now with https://github.com/latex-lsp/texlab/commit/ae4a1266b914b8980346277ff51ad7dc2f31c6e3. The new approach should no longer cause race conditions nor cause some LSP clients to hang sometimes.

**Comment by pfoerster** [comment]
This fix has been released with `texlab` 3.3.2.
